// aadhar-charity-platform/backend/controllers/notificationController.js
const Notification = require("../models/Notification");
const User = require("../models/User");
const webpush = require("web-push");

if (process.env.VAPID_EMAIL && process.env.VAPID_PUBLIC_KEY && process.env.VAPID_PRIVATE_KEY) {
  webpush.setVapidDetails(
    process.env.VAPID_EMAIL,
    process.env.VAPID_PUBLIC_KEY,
    process.env.VAPID_PRIVATE_KEY
  );
} else {
  console.log('⚠️ WebPush VAPID not configured');
}

// @desc    Get VAPID public key
// @route   GET /api/notifications/vapid-key
// @access  Public
exports.getVapidKey = (req, res) => {
  res.json({ success: true, publicKey: process.env.VAPID_PUBLIC_KEY });
};

// @desc    Subscribe to push notifications
// @route   POST /api/notifications/subscribe
// @access  Private
exports.subscribe = async (req, res, next) => {
  try {
    const { subscription } = req.body;

    if (!subscription || !subscription.endpoint) {
      return res.status(400).json({ success: false, error: "Invalid subscription object" });
    }

    const user = await User.findById(req.user.id);

    // Check if already subscribed with this endpoint
    const exists = user.pushSubscriptions.some((s) => s.endpoint === subscription.endpoint);

    if (!exists) {
      user.pushSubscriptions.push(subscription);
      await user.save();
    }

    // Send a welcome notification
    try {
      await webpush.sendNotification(
        subscription,
        JSON.stringify({
          title: "AADHAR Notifications Active 🔔",
          body: "You'll now receive alerts for urgent causes near you.",
          icon: "/icons/icon-192x192.png",
        })
      );
    } catch (e) {
      // ignore welcome notification failures
    }

    res.json({ success: true, message: "Push notification subscription saved" });
  } catch (err) {
    next(err);
  }
};

// @desc    Unsubscribe from push notifications
// @route   DELETE /api/notifications/subscribe
// @access  Private
exports.unsubscribe = async (req, res, next) => {
  try {
    const { endpoint } = req.body;
    const user = await User.findById(req.user.id);
    user.pushSubscriptions = user.pushSubscriptions.filter((s) => s.endpoint !== endpoint);
    await user.save();

    res.json({ success: true, message: "Unsubscribed from push notifications" });
  } catch (err) {
    next(err);
  }
};

// @desc    Get user notifications
// @route   GET /api/notifications
// @access  Private
exports.getNotifications = async (req, res, next) => {
  try {
    const { page = 1, limit = 20, unreadOnly } = req.query;
    const query = { recipient: req.user.id };
    if (unreadOnly === "true") query.isRead = false;

    const notifications = await Notification.find(query)
      .sort({ createdAt: -1 })
      .skip((page - 1) * limit)
      .limit(parseInt(limit));

    const unreadCount = await Notification.countDocuments({ recipient: req.user.id, isRead: false });

    res.json({
      success: true,
      unreadCount,
      data: notifications,
    });
  } catch (err) {
    next(err);
  }
};

// @desc    Mark notification(s) as read
// @route   PATCH /api/notifications/read
// @access  Private
exports.markAsRead = async (req, res, next) => {
  try {
    const { ids, markAll } = req.body;

    if (markAll) {
      await Notification.updateMany({ recipient: req.user.id, isRead: false }, { isRead: true, readAt: new Date() });
    } else if (ids && ids.length) {
      await Notification.updateMany(
        { _id: { $in: ids }, recipient: req.user.id },
        { isRead: true, readAt: new Date() }
      );
    }

    res.json({ success: true, message: "Notifications marked as read" });
  } catch (err) {
    next(err);
  }
};

// @desc    Delete notification
// @route   DELETE /api/notifications/:id
// @access  Private
exports.deleteNotification = async (req, res, next) => {
  try {
    await Notification.findOneAndDelete({ _id: req.params.id, recipient: req.user.id });
    res.json({ success: true, message: "Notification deleted" });
  } catch (err) {
    next(err);
  }
};
