// aadhar-charity-platform/backend/models/User.js
const mongoose = require("mongoose");
const bcrypt = require("bcryptjs");
const jwt = require("jsonwebtoken");

const UserSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: [true, "Name is required"],
      trim: true,
      maxlength: [50, "Name cannot exceed 50 characters"],
    },
    email: {
      type: String,
      required: [true, "Email is required"],
      unique: true,
      lowercase: true,
      match: [/^\S+@\S+\.\S+$/, "Please enter a valid email"],
    },
    password: {
      type: String,
      required: [true, "Password is required"],
      minlength: [6, "Password must be at least 6 characters"],
      select: false,
    },
    role: {
      type: String,
      enum: ["donor", "charity", "admin"],
      default: "donor",
    },
    avatar: {
      type: String,
      default: "",
    },
    phone: {
      type: String,
      default: "",
    },
    // GeoJSON location for KNN
    location: {
      type: {
        type: String,
        enum: ["Point"],
        default: "Point",
      },
      coordinates: {
        type: [Number], // [longitude, latitude]
        default: [0, 0],
      },
      address: { type: String, default: "" },
      city: { type: String, default: "" },
      state: { type: String, default: "" },
      pincode: { type: String, default: "" },
    },
    // KNN Preference tracking
    preferences: {
      categories: {
        type: [String],
        enum: [
          "Education",
          "Healthcare",
          "Poverty",
          "Environment",
          "Animal Welfare",
          "Disaster Relief",
          "Women Empowerment",
          "Child Welfare",
          "Elderly Care",
          "Disability Support",
          "Arts & Culture",
          "Sports",
          "Water & Sanitation",
          "Food Security",
          "Rural Development",
        ],
        default: [],
      },
      // Click weights for KNN personalization
      categoryWeights: {
        type: Map,
        of: Number,
        default: {},
      },
      maxDistanceKm: {
        type: Number,
        default: 50,
      },
    },
    // Followed charities
    followedCharities: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "Charity",
      },
    ],
    // Push notification subscription
    pushSubscriptions: [
      {
        endpoint: String,
        keys: {
          p256dh: String,
          auth: String,
        },
        createdAt: { type: Date, default: Date.now },
      },
    ],
    fcmTokens: [{
      token: String,
      createdAt: { type: Date, default: Date.now }
    }],
    // Interaction history for KNN
    firebaseUid: {
      type: String,
      sparse: true,
      unique: true,
    },
    interactionHistory: [
      {
        category: String,
        charityId: { type: mongoose.Schema.Types.ObjectId, ref: "Charity" },
        action: { type: String, enum: ["view", "click", "donate", "follow"] },
        weight: { type: Number, default: 1 },
        timestamp: { type: Date, default: Date.now },
      },
    ],
    isVerified: { type: Boolean, default: false },
    otp: {
      type: String,
      select: false
    },
    otpExpiry: {
      type: Date,
      select: false
    },
    isActive: { type: Boolean, default: true },
    lastLogin: Date,
    passwordResetToken: String,
    passwordResetExpire: Date,
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Geo index for KNN queries
UserSchema.index({ "location.coordinates": "2dsphere" });

// Hash password before save
UserSchema.pre("save", async function (next) {
  if (!this.isModified("password")) return next();
  const salt = await bcrypt.genSalt(12);
  this.password = await bcrypt.hash(this.password, salt);
  next();
});

// Sign JWT
UserSchema.methods.getSignedJwtToken = function () {
  return jwt.sign({ id: this._id, role: this.role }, process.env.JWT_SECRET, {
    expiresIn: process.env.JWT_EXPIRE,
  });
};

// Match password
UserSchema.methods.matchPassword = async function (enteredPassword) {
  return await bcrypt.compare(enteredPassword, this.password);
};

// Update category weight for KNN
UserSchema.methods.updateCategoryWeight = async function (category, action) {
  const weights = { view: 1, click: 2, donate: 5, follow: 3 };
  const currentWeight = this.preferences.categoryWeights.get(category) || 0;
  this.preferences.categoryWeights.set(category, currentWeight + (weights[action] || 1));
  await this.save();
};

module.exports = mongoose.model("User", UserSchema);
